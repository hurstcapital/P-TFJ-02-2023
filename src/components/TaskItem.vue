<template>
  <div
    class="flex justify-between items-center bg-white rounded-md shadow-md my-2 p-1 border-4"
    :class="isComplete ? 'border-green-500' : 'border-white'"
    data-cy="task-item"
  >
    <div
      class="flex items-center"
      @click="toggleCompletion()"
      @keydown.enter="toggleCompletion()"
    >
      <div class="flex-shrink-0 ml-3 align-middle">
        <BaseInputCheckBox :isChecked="isComplete" />
      </div>
      <div class="ml-2">
        <h4
          class="text-xl text-gray-900 leading-tight"
          :class="isComplete ? 'line-through' : ''"
        >
          {{ description }}
        </h4>
      </div>
    </div>
    <div class="flex items-center">
      <BaseButton @click="removeTask"> Deletar </BaseButton>
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent } from "vue";
import { useStore } from "@/store";
import { IMutationType } from "@/interfaces/Mutations";
import BaseButton from "@/components/common/BaseButton.vue";
import BaseInputCheckBox from "@/components/common/BaseInputCheckBox.vue";

export default defineComponent({
  name: "TaskItem",
  components: {
    BaseButton,
    BaseInputCheckBox,
  },
  props: {
    id: { type: String, required: true },
    description: { type: String, required: true },
    isComplete: { type: Boolean, required: true },
  },
  setup(props) {
    const store = useStore();

    const toggleCompletion = () => {
      store.commit(IMutationType.CompleteItem, {
        id: props.id,
        isComplete: !props.isComplete,
      });
    };

    const removeTask = () => {
      store.commit(IMutationType.RemoveTask, { id: props.id });
    };

    return { toggleCompletion, removeTask };
  },
});
</script>
