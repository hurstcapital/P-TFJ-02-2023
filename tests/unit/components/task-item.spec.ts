import { shallowMount } from "@vue/test-utils";
import TaskItem from "@/components/TaskItem.vue";

const description = "description example";
const id = "123";

const optionsWrapper = (isComplete = false) => {
  return {
    stubs: ["baseinput-check-box", "base-button"],
    props: {
      description: description,
      id: id,
      isComplete: isComplete,
    },
  };
};

describe("TaskItem.vue", () => {
  it("should renders a description", () => {
    const description = "description example";

    const wrapper = shallowMount(TaskItem, {
      ...optionsWrapper(),
    });

    expect(wrapper.find("h4").text()).toMatch(description);
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should be completed", async () => {
    const wrapper = shallowMount(TaskItem, {
      ...optionsWrapper(true),
    });

    expect(wrapper.find("h4").classes()).toContain("line-through");
    expect(wrapper.find(".border-green-500").exists()).toBeTruthy();

    expect(wrapper.element).toMatchSnapshot();
  });
});
