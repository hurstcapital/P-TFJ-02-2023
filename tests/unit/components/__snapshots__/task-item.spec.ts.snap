// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskItem.vue should be completed 1`] = `
<div
  class="flex justify-between items-center bg-white rounded-md shadow-md my-2 p-1 border-4 border-green-500"
  data-cy="task-item"
>
  <div
    class="flex items-center"
  >
    <div
      class="flex-shrink-0 ml-3 align-middle"
    >
      <base-input-check-box-stub
        ischecked="true"
      />
    </div>
    <div
      class="ml-2"
    >
      <h4
        class="text-xl text-gray-900 leading-tight line-through"
      >
        description example
      </h4>
    </div>
  </div>
  <div
    class="flex items-center"
  >
    <base-button-stub
      isdisabled="false"
    />
  </div>
</div>
`;

exports[`TaskItem.vue should renders a description 1`] = `
<div
  class="flex justify-between items-center bg-white rounded-md shadow-md my-2 p-1 border-4 border-white"
  data-cy="task-item"
>
  <div
    class="flex items-center"
  >
    <div
      class="flex-shrink-0 ml-3 align-middle"
    >
      <base-input-check-box-stub
        ischecked="false"
      />
    </div>
    <div
      class="ml-2"
    >
      <h4
        class="text-xl text-gray-900 leading-tight"
      >
        description example
      </h4>
    </div>
  </div>
  <div
    class="flex items-center"
  >
    <base-button-stub
      isdisabled="false"
    />
  </div>
</div>
`;
